File: app/src/main/java/com/pec_acm/moviedroid/MainActivity.java
package com.pec_acm.moviedroid;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private static final String TAG = "MainActivity";
    private static final String ARG_NAME = "username";
    public static void startActivity(Context context, String username) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.putExtra(ARG_NAME, username);
        context.startActivity(intent);
    }
    FirebaseAuth firebaseAuth;
    GoogleSignInClient googleSignInClient;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        TextView textView = findViewById(R.id.textViewWelcome);
        if (getIntent().hasExtra(ARG_NAME)) {
            textView.setText(String.format("Welcome - %s", getIntent().getStringExtra(ARG_NAME)));
        }
        findViewById(R.id.buttonLogout).setOnClickListener(this);
        findViewById(R.id.buttonDisconnect).setOnClickListener(this);
        googleSignInClient = GoogleSignIn.getClient(this, GoogleSignInOptions.DEFAULT_SIGN_IN);
        firebaseAuth = FirebaseAuth.getInstance();
    }
    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.buttonLogout:
                signOut();
                break;
            case R.id.buttonDisconnect:
                revokeAccess();
                break;
        }
    }
    private void signOut() {
        // Firebase sign out
        firebaseAuth.signOut();
        // Google sign out
        googleSignInClient.signOut().addOnCompleteListener(this,
                new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        // Google Sign In failed, update UI appropriately
                        Log.w(TAG, "Signed out of google");
                    }
                });
    }
    private void revokeAccess() {
        // Firebase sign out
        firebaseAuth.signOut();
        // Google revoke access
        googleSignInClient.revokeAccess().addOnCompleteListener(this,
                new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        // Google Sign In failed, update UI appropriately
                        Log.w(TAG, "Revoked Access");
                    }
                });
    }
}

File: app/src/main/java/com/pec_acm/moviedroid/MainActivity.kt
package com.pec_acm.moviedroid


import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.findNavController
import androidx.navigation.ui.AppBarConfiguration
import androidx.navigation.ui.setupWithNavController
import com.bumptech.glide.Glide
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.auth.FirebaseAuth
import com.pec_acm.moviedroid.databinding.ActivityMainBinding
import com.pec_acm.moviedroid.mainpage.list.ListViewModel
import com.pec_acm.moviedroid.mainpage.search.SearchFragment
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private lateinit var authListener: FirebaseAuth.AuthStateListener
    private val logoutObserver: MutableLiveData<FirebaseAuth> = MutableLiveData()
    private var isBottomNavHidden = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)


        //Implementing Bottom Navigation View
        val navController = findNavController(R.id.main_page_fragment_container)

        binding.bottomNavBar.setupWithNavController(navController)

        val bottomNavDestinations = setOf(
            R.id.homeFragment,
            R.id.searchFragment,
            R.id.listFragment
        )

        // Implementing Navigation drawer
        val appBarConfiguration = AppBarConfiguration(
            bottomNavDestinations,
            binding.dlParent
        )
        binding.toolBar.setupWithNavController(navController, appBarConfiguration)

        val listViewModel = ViewModelProvider(this)[ListViewModel::class.java]
        val user = FirebaseAuth.getInstance().currentUser

        listViewModel.addUser(user!!.uid, user.displayName, user.photoUrl?.toString(), user.email)
        listViewModel.getUser(user.uid)

        val headerView: View = binding.sideNavBar.getHeaderView(0)

        val tvUserName: TextView = headerView.findViewById(R.id.tv_user_name)
        val tvEmailId: TextView = headerView.findViewById(R.id.tv_email_id)
        val ivProfilePic: ImageView = headerView.findViewById(R.id.iv_profile_picture)

        tvUserName.text = user.displayName
        tvEmailId.text = user.email

        Glide.with(this)
            .load(user.photoUrl)
            .placeholder(R.drawable.ic_baseline_account_circle_24)
            .into(ivProfilePic)

        navController.addOnDestinationChangedListener { _, destination, _ ->
            if (destination.id in bottomNavDestinations) {
                showBottomNavigation()
                showToolBar()
            } else {
                binding.bottomNavBar.visibility = View.GONE
                binding.toolBar.visibility = View.GONE
            }
        }

        binding.sideNavBar.setNavigationItemSelectedListener {
            when (it.itemId) {
                R.id.item_show_logout_dialog -> {
                    navController.navigate(R.id.show_log_out_dialog)
                }

                R.id.item_move_to_profile_page -> {
                    //move to profile activity
                    val intent = Intent(this, ProfileActivity::class.java)
                    startActivity(intent)
                }
                else -> {
                    Snackbar.make(
                        binding.root,
                        R.string.some_awesome_feature_is_on_its_way,
                        Snackbar.LENGTH_SHORT
                    ).show()
                }
            }

            binding.dlParent.close()
            return@setNavigationItemSelectedListener true
        }

        authListener = FirebaseAuth.AuthStateListener { firebaseAuth ->
            if (null == firebaseAuth.currentUser) {
                logoutObserver.value = firebaseAuth
            }
        }

        FirebaseAuth.getInstance().addAuthStateListener(authListener)

        logoutObserver.observe(this) {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
        }
    }

    fun hideBottomNavigation() {
        binding.bottomNavBar
            .animate()
            .translationY(binding.bottomNavBar.height.toFloat())
            .setDuration(300)

    }

    fun showBottomNavigation() {
        binding.bottomNavBar.visibility = View.VISIBLE
        binding.bottomNavBar
            .animate()
            .translationY(0f)
            .setDuration(300)
    }

    fun showToolBar() {
        binding.bottomNavBar.visibility = View.VISIBLE
        binding.toolBar.visibility = View.VISIBLE
    }

    override fun onDestroy() {
        super.onDestroy()

        FirebaseAuth.getInstance().removeAuthStateListener(authListener)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        return when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                true
            }
            else -> {
                super.onOptionsItemSelected(item)
            }
        }
    }

    override fun onBackPressed() {
        val navHost = supportFragmentManager.findFragmentById(R.id.main_page_fragment_container)
        val fragment = navHost?.childFragmentManager?.primaryNavigationFragment
        when (fragment) {
            is SearchFragment -> {
                if (fragment.searchText.isIconified) {
                    super.onBackPressed()
                } else {
                    fragment.searchText.isIconified = true
                }
            }
            else -> {
                super.onBackPressed()
            }
        }
    }
}